apypie/foreman.py

```diff
@@ -0,0 +1,227 @@
+        api_action = self._resource(resource).action(action)
+        return api_action.prepare_params(params)
+
+    def resource_action(self, resource: str, action: str, params: dict, options=None, data=None, files=None,  # pylint: disable=too-many-arguments
+                        ignore_task_errors: bool = False) -> Optional[dict]:
+        """
+        Perform a generic action on a resource
+
+        Will wait for tasks if the action returns one
+        """
+        resource_payload = self._resource_prepare_params(resource, action, params)
+        if options is None:
+            options = {}
+        try:
+            result = self._resource_call(resource, action, resource_payload, options=options, data=data, files=files)
+            is_foreman_task = isinstance(result, dict) and 'action' in result and 'state' in result and 'started_at' in result
+            if result and is_foreman_task:
+                result = self.wait_for_task(result, ignore_errors=ignore_task_errors)
+        except Exception as exc:
+            msg = f'Error while performing {action} on {resource}: {exc}'
+            raise ForemanApiException.from_exception(exc, msg) from exc
+        return result
+
+    def wait_for_task(self, task: dict, ignore_errors: bool = False) -> dict:
+        """
+        Wait for a foreman-tasks task, polling it every ``self.task_poll`` seconds.
+
+        Will raise a ForemanApiException when task has not finished in ``self.task_timeout`` seconds.
+        """
+        duration = self.task_timeout
+        while task['state'] not in ['paused', 'stopped']:
+            duration -= self.task_poll
+            if duration <= 0:
+                raise ForemanApiException(msg=f"Timeout waiting for Task {task['id']}")
+            time.sleep(self.task_poll)
+
+            resource_payload = self._resource_prepare_params('foreman_tasks', 'show', {'id': task['id']})
+            task = cast(dict, self._resource_call('foreman_tasks', 'show', resource_payload))
+        if not ignore_errors and task['result'] != 'success':
+            msg = f"Task {task['action']}({task['id']}) did not succeed. Task information: {task['humanized']['errors']}"
+            raise ForemanApiException(msg=msg)
+        return task
+
+    def show(self, resource: str, resource_id: int, params: Optional[dict] = None) -> Optional[dict]:
+        """
+        Execute the ``show`` action on an entity.
+
+        :param resource: Plural name of the api resource to show
+        :param resource_id: The ID of the entity to show
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+
+        :return: The entity
+        """
+        payload = {'id': resource_id}
+        if params:
+            payload.update(params)
+        return self.resource_action(resource, 'show', payload)
+
+    def list(self, resource: str, search: Optional[str] = None, params: Optional[dict] = None) -> list:
```

Comment:
should this be called `list` (as that's the natural thing it returns) or `index` as that's the action Foreman defines?

Reply:
The benefit of `index` is that it doesn't overlap with the built in `list()` method, but it feels less natural to me.